-Ainfer causes crash with multi-dimensional arrays

When I use the `-Ainfer` option on the following code, I code a crash. It seems to happen with two-dimensional arrays.
```java
public class C {
    boolean[][] field = null;
}
```
```bash
$CHECKERFRAMEWORK/checker/bin/javac -processor org.checkerframework.checker.nullness.NullnessChecker -Ainfer=stubs C.java
```
```
error: SourceChecker.typeProcess: unexpected Throwable (NullPointerException) while processing C.java
  ; The Checker Framework crashed.  Please report the crash.
  Compilation unit: C.java
  Last visited tree at line 2 column 25:
      boolean[][] field = null;
  Exception: java.lang.NullPointerException; java.lang.NullPointerException
  	at org.checkerframework.common.wholeprograminference.SceneToStubWriter.formatArrayTypeImpl(SceneToStubWriter.java:244)
  	at org.checkerframework.common.wholeprograminference.SceneToStubWriter.formatArrayType(SceneToStubWriter.java:179)
  	at org.checkerframework.common.wholeprograminference.SceneToStubWriter.formatType(SceneToStubWriter.java:384)
  	at org.checkerframework.common.wholeprograminference.SceneToStubWriter.formatType(SceneToStubWriter.java:347)
  	at org.checkerframework.common.wholeprograminference.SceneToStubWriter.formatAFieldImpl(SceneToStubWriter.java:315)
  	at org.checkerframework.common.wholeprograminference.SceneToStubWriter.printField(SceneToStubWriter.java:514)
  	at org.checkerframework.common.wholeprograminference.SceneToStubWriter.printFields(SceneToStubWriter.java:493)
  	at org.checkerframework.common.wholeprograminference.SceneToStubWriter.printClass(SceneToStubWriter.java:693)
  	at org.checkerframework.common.wholeprograminference.SceneToStubWriter.writeImpl(SceneToStubWriter.java:616)
  	at org.checkerframework.common.wholeprograminference.SceneToStubWriter.write(SceneToStubWriter.java:89)
  	at org.checkerframework.common.wholeprograminference.scenelib.ASceneWrapper.writeToFile(ASceneWrapper.java:146)
  	at org.checkerframework.common.wholeprograminference.WholeProgramInferenceScenesStorage.writeScenes(WholeProgramInferenceScenesStorage.java:110)
  	at org.checkerframework.common.wholeprograminference.WholeProgramInferenceScenes.writeResultsToFile(WholeProgramInferenceScenes.java:423)
  	at org.checkerframework.framework.type.AnnotatedTypeFactory.postProcessClassTree(AnnotatedTypeFactory.java:1121)
  	at org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:327)
  	at org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:165)
  	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:808)
  	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
  	at org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:82)
  	at org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:894)
  	at org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:510)
  	at org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:190)
  	at jdk.compiler/com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:828)
  	at jdk.compiler/com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:132)
  	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1418)
  	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1375)
  	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:973)
  	at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:311)
  	at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:170)
  	at jdk.compiler/com.sun.tools.javac.Main.compile(Main.java:57)
  	at jdk.compiler/com.sun.tools.javac.Main.main(Main.java:43)
C.java:2: error: [assignment.type.incompatible] incompatible types in assignment.
    boolean[][] field = null;
                        ^
  found   : null
  required: @Initialized @NonNull boolean @Initialized @NonNull [] @Initialized @NonNull []
2 errors
```

When I use `bin-devel/javac`, I get a different crash.
```bash
$CHECKERFRAMEWORK/checker/bin-devel/javac -processor org.checkerframework.checker.nullness.NullnessChecker -Ainfer=stubs C.java
```
```
error: SourceChecker.typeProcess: unexpected Throwable (NoSuchMethodError) while processing C.java; message: 'void scenelib.annotations.el.AClass.setTypeElement(javax.lang.model.element.TypeElement)'
  ; The Checker Framework crashed.  Please report the crash.
  Compilation unit: C.java
  Last visited tree at line 1 column 1:
  public class C {
  Exception: java.lang.NoSuchMethodError: 'void scenelib.annotations.el.AClass.setTypeElement(javax.lang.model.element.TypeElement)'; java.lang.NoSuchMethodError: 'void scenelib.annotations.el.AClass.setTypeElement(javax.lang.model.element.TypeElement)'
  	at org.checkerframework.common.wholeprograminference.scenelib.ASceneWrapper.updateSymbolInformation(ASceneWrapper.java:219)
  	at org.checkerframework.common.wholeprograminference.WholeProgramInferenceScenesStorage.getAClass(WholeProgramInferenceScenesStorage.java:170)
  	at org.checkerframework.common.wholeprograminference.WholeProgramInferenceScenes.updateFromFieldAssignment(WholeProgramInferenceScenes.java:302)
  	at org.checkerframework.framework.flow.CFAbstractTransfer.visitAssignment(CFAbstractTransfer.java:814)
  	at org.checkerframework.framework.flow.CFAbstractTransfer.visitAssignment(CFAbstractTransfer.java:91)
  	at org.checkerframework.dataflow.cfg.node.AssignmentNode.accept(AssignmentNode.java:62)
  	at org.checkerframework.dataflow.analysis.Analysis.callTransferFunction(Analysis.java:437)
  	at org.checkerframework.dataflow.analysis.Analysis.performAnalysisBlock(Analysis.java:226)
  	at org.checkerframework.dataflow.analysis.Analysis.performAnalysis(Analysis.java:201)
  	at org.checkerframework.framework.flow.CFAbstractAnalysis.performAnalysis(CFAbstractAnalysis.java:109)
  	at org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:1243)
  	at org.checkerframework.framework.type.GenericAnnotatedTypeFactory.performFlowAnalysis(GenericAnnotatedTypeFactory.java:1111)
  	at org.checkerframework.framework.type.GenericAnnotatedTypeFactory.checkAndPerformFlowAnalysis(GenericAnnotatedTypeFactory.java:1527)
  	at org.checkerframework.framework.type.GenericAnnotatedTypeFactory.preProcessClassTree(GenericAnnotatedTypeFactory.java:258)
  	at org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:307)
  	at org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:165)
  	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:808)
  	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
  	at org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:82)
  	at org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:894)
  	at org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:510)
  	at org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:503)
  	at org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:190)
  	at jdk.compiler/com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:828)
  	at jdk.compiler/com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:132)
  	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1418)
  	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1375)
  	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:973)
  	at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:311)
  	at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:170)
  	at jdk.compiler/com.sun.tools.javac.Main.compile(Main.java:57)
  	at jdk.compiler/com.sun.tools.javac.Main.main(Main.java:43)
error: SourceChecker.typeProcess: unexpected Throwable (NoSuchMethodError) while processing C.java; message: 'void scenelib.annotations.el.AClass.setTypeElement(javax.lang.model.element.TypeElement)'
  ; The Checker Framework crashed.  Please report the crash.
  Compilation unit: C.java
  Last visited tree at line 1 column 1:
  public class C {
  Exception: java.lang.NoSuchMethodError: 'void scenelib.annotations.el.AClass.setTypeElement(javax.lang.model.element.TypeElement)'; java.lang.NoSuchMethodError: 'void scenelib.annotations.el.AClass.setTypeElement(javax.lang.model.element.TypeElement)'
  	at org.checkerframework.common.wholeprograminference.scenelib.ASceneWrapper.updateSymbolInformation(ASceneWrapper.java:219)
  	at org.checkerframework.common.wholeprograminference.WholeProgramInferenceScenesStorage.getAClass(WholeProgramInferenceScenesStorage.java:170)
  	at org.checkerframework.common.wholeprograminference.WholeProgramInferenceScenes.updateFromFieldAssignment(WholeProgramInferenceScenes.java:302)
  	at org.checkerframework.framework.flow.CFAbstractTransfer.visitAssignment(CFAbstractTransfer.java:814)
  	at org.checkerframework.checker.initialization.InitializationTransfer.visitAssignment(InitializationTransfer.java:149)
  	at org.checkerframework.checker.initialization.InitializationTransfer.visitAssignment(InitializationTransfer.java:57)
  	at org.checkerframework.dataflow.cfg.node.AssignmentNode.accept(AssignmentNode.java:62)
  	at org.checkerframework.dataflow.analysis.Analysis.callTransferFunction(Analysis.java:437)
  	at org.checkerframework.dataflow.analysis.Analysis.performAnalysisBlock(Analysis.java:226)
  	at org.checkerframework.dataflow.analysis.Analysis.performAnalysis(Analysis.java:201)
  	at org.checkerframework.framework.flow.CFAbstractAnalysis.performAnalysis(CFAbstractAnalysis.java:109)
  	at org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:1243)
  	at org.checkerframework.framework.type.GenericAnnotatedTypeFactory.performFlowAnalysis(GenericAnnotatedTypeFactory.java:1111)
  	at org.checkerframework.framework.type.GenericAnnotatedTypeFactory.checkAndPerformFlowAnalysis(GenericAnnotatedTypeFactory.java:1527)
  	at org.checkerframework.framework.type.GenericAnnotatedTypeFactory.preProcessClassTree(GenericAnnotatedTypeFactory.java:258)
  	at org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:307)
  	at org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:165)
  	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:808)
  	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
  	at org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:82)
  	at org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:894)
  	at org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:510)
  	at org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:190)
  	at jdk.compiler/com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:828)
  	at jdk.compiler/com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:132)
  	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1418)
  	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1375)
  	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:973)
  	at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:311)
  	at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:170)
  	at jdk.compiler/com.sun.tools.javac.Main.compile(Main.java:57)
  	at jdk.compiler/com.sun.tools.javac.Main.main(Main.java:43)
2 errors
```

Here's the output for
```bash
$CHECKERFRAMEWORK/checker/bin/javac -processor org.checkerframework.checker.nullness.NullnessChecker -Ainfer=stubs C.java -version -verbose -AprintAllQualifiers
```
and
```bash
$CHECKERFRAMEWORK/checker/bin-devel/javac -processor org.checkerframework.checker.nullness.NullnessChecker -Ainfer=stubs C.java -version -verbose -AprintAllQualifiers
```

[bin-javac-output.txt](https://github.com/typetools/checker-framework/files/4868687/bin-javac-output.txt)
[bin-devel-javac-output.txt](https://github.com/typetools/checker-framework/files/4868690/bin-devel-javac-output.txt)

--- a/framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java
+++ b/framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java
@@ -200,9 +200,12 @@ public final class SceneToStubWriter {
      * @param levels the number of component types the type should have, derived from the javac
      *     representation
      * @return a list of the array levels in scenelib's representation, but in the order used by
-     *     javac. Guaranteed to have exactly {@code levels} entries.
+     *     javac. Guaranteed to have exactly {@code levels} entries. Entries may be null, if the
+     *     corresponding parts of {@code scenelibRep} are null. See <a
+     *     href="https://github.com/typetools/checker-framework/issues/3422">issue 3422</a> for an
+     *     example of code that causes a null ATypeElement.
      */
-    private static List<ATypeElement> getSceneLibRepInJavacOrder(
+    private static List<@Nullable ATypeElement> getSceneLibRepInJavacOrder(
             ATypeElement scenelibRep, int levels) {
         List<ATypeElement> result = new ArrayList<>();
         ATypeElement array = scenelibRep;
@@ -225,12 +228,13 @@ public final class SceneToStubWriter {
      * using {@link #formatType(ATypeElement, TypeMirror)}.
      *
      * @param scenelibRepInJavacOrder the scenelib representation, reordered to match javac's order.
-     *     See {@link #getSceneLibRepInJavacOrder} for an explanation of why this is necessary.
+     *     See {@link #getSceneLibRepInJavacOrder} for an explanation of why this is necessary and
+     *     why the elements may be null.
      * @param javacRep the javac representation of the array type
      * @return the type formatted to be written to Java source code, followed by a space character
      */
     private static String formatArrayTypeImpl(
-            List<ATypeElement> scenelibRepInJavacOrder, ArrayType javacRep) {
+            List<@Nullable ATypeElement> scenelibRepInJavacOrder, ArrayType javacRep) {
         TypeMirror javacComponent = javacRep.getComponentType();
         ATypeElement scenelibRep = scenelibRepInJavacOrder.get(0);
         ATypeElement scenelibComponent = scenelibRepInJavacOrder.get(1);
@@ -241,7 +245,9 @@ public final class SceneToStubWriter {
             result += " ";
         }
         if ("".equals(result)) {
-            result += formatAnnotations(scenelibRep.tlAnnotationsHere);
+            if (scenelibRep != null) {
+                result += formatAnnotations(scenelibRep.tlAnnotationsHere);
+            }
         }
         result += "[] ";
         if (javacComponent.getKind() == TypeKind.ARRAY) {

--- a/framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java
+++ b/framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java
@@ -203,7 +203,8 @@ public final class SceneToStubWriter {
      *     javac. Guaranteed to have exactly {@code levels} entries. Entries may be null, if the
      *     corresponding parts of {@code scenelibRep} are null. See <a
      *     href="https://github.com/typetools/checker-framework/issues/3422">issue 3422</a> for an
-     *     example of code that causes a null ATypeElement.
+     *     example of code that causes a null ATypeElement, because the component type is unknown,
+     *     but the primary type of the array is known.
      */
     private static List<@Nullable ATypeElement> getSceneLibRepInJavacOrder(
             ATypeElement scenelibRep, int levels) {
@@ -244,10 +245,8 @@ public final class SceneToStubWriter {
             result += explicitAnno.toString();
             result += " ";
         }
-        if ("".equals(result)) {
-            if (scenelibRep != null) {
-                result += formatAnnotations(scenelibRep.tlAnnotationsHere);
-            }
+        if (result.isEmpty() && scenelibRep != null) {
+            result += formatAnnotations(scenelibRep.tlAnnotationsHere);
         }
         result += "[] ";
         if (javacComponent.getKind() == TypeKind.ARRAY) {

--- a/framework/tests/whole-program-inference/non-annotated/MultidimensionalAnnotatedArray.java
+++ b/framework/tests/whole-program-inference/non-annotated/MultidimensionalAnnotatedArray.java
@@ -1,0 +1,11 @@
+// test case for https://github.com/typetools/checker-framework/issues/3422
+
+import testlib.wholeprograminference.qual.Sibling1;
+
+public class MultidimensionalAnnotatedArray {
+    boolean[][] field = getArray();
+
+    public boolean[] @Sibling1 [] getArray() {
+        return null;
+    }
+}

